### PR TITLE
internal/v4: implement revision-info and expand-id

### PR DESCRIPTION
**Pull request by rogpeppe**
This involves exporting a few more things from v5 too, and
adding some more tests.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1374/
Test PASSed.

File: internal/v4/api.go [hunk 1]
```diff
@@ -5,17 +5,25 @@ package v4 // import "gopkg.in/juju/charmstore.v5-unstable/internal/v4"
...
+	"gopkg.in/mgo.v2/bson"
 
 	"gopkg.in/juju/charmstore.v5-unstable/internal/charmstore"
 	"gopkg.in/juju/charmstore.v5-unstable/internal/mempool"
+	"gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
 	"gopkg.in/juju/charmstore.v5-unstable/internal/router"
 	"gopkg.in/juju/charmstore.v5-unstable/internal/v5"
 )
 
+var logger = loggo.GetLogger("charmstore.internal.v5")
```
Comment (mhilton) [inline internal/v4/api.go]: v4

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1375/
Test PASSed.

File: internal/v4/api.go [hunk 2]
```diff
@@ -152,3 +163,105 @@ func StatsEnabled(req *http.Request) bool {
...
+	} else {
+		q = q.Sort("-series", "-revision")
+	}
+	var docs []*mongodoc.Entity
+	err := q.All(&docs)
+	if err != nil && errgo.Cause(err) != mgo.ErrNotFound {
+		return errgo.Mask(err)
+	}
+
+	// A not found error should have been already returned by the router in the
```
Comment (mhilton) [inline internal/v4/api.go]: I believe this code is now superfluous as ResolveURL will always return a not found error when the entity does not exist.

File: internal/v4/api.go [hunk 3]
```diff
@@ -152,3 +163,105 @@ func StatsEnabled(req *http.Request) bool {
...
+		url := doc.PreferredURL(id.PromulgatedRevision != -1)
+		url.Series = series
+		response = append(response, params.ExpandedId{Id: url.String()})
+	})
+
+	// Write the response in JSON format.
+	return httprequest.WriteJSON(w, http.StatusOK, response)
+}
+
+func expandMultiSeries(entities []*mongodoc.Entity, append func(series string, doc *mongodoc.Entity)) {
```
Comment (mhilton) [inline internal/v4/api.go:256]: Some godoc here would be nice.

**Comment by mhilton** [comment]
:+1: with some small comments.

File: internal/v4/api.go (hunk 2 above)
Comment (mhilton) [inline internal/v4/api.go]: This is probably true in v5 as well.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1376/
Test PASSed.

File: internal/v4/api.go (hunk 2 above)
Comment (rogpeppe) [inline internal/v4/api.go]: removed

File: internal/v4/api.go [hunk 4]
```diff
@@ -152,3 +163,104 @@ func StatsEnabled(req *http.Request) bool {
...
+		q = q.Sort("-series", "-promulgated-revision")
+	} else {
+		q = q.Sort("-series", "-revision")
+	}
+	var docs []*mongodoc.Entity
+	err := q.All(&docs)
+	if err != nil && errgo.Cause(err) != mgo.ErrNotFound {
+		return errgo.Mask(err)
+	}
+
```
Comment (fabricematrat) [inline internal/v4/api.go:236]: the query seems the same than metaRevisionInfo so may be can be extracted ?

**Comment by fabricematrat** [comment]
:+1:

File: internal/v4/api.go (hunk 4 above)
Comment (rogpeppe) [inline internal/v4/api.go:236]: Not similar enough, I think.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**